language/move-vm/types/src/values/mod.rs

```diff
@@ -1,6 +1,18 @@
 // Copyright (c) The Diem Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
+//! Implementation for Move values
+//!
+//! REFERENCE SAFETY EXPLANATION
+//! References in Move are proved safe statically in the reference safety check
+//! (bytecode_verifier/reference_safety). This check guarantees that there are no dangling
+//! references and no memory leaks. But the access pattern for reading/writing to Values must
+//! abide by these safety rules (which the VM interpreter does. You should refrain from writing to
+//! reference values outside of the VM or native functions (with special care))
+//! However, it is **very important** that any instance of `Locals` is not dropped until after
+//! all Values are consumed. If this is not followed, reading/writing to Values *will* read from
+//! invalid pointers.
+
```

Comment:
I think this is kind of bummer for usage of the VM. We want to use references in adapters (doesn't Sui does so) and custom embeddings of the VM, as well as native functions implementations, avoiding the need for using `unsafe` all over the code. Unsafe Rust should be isolated. Can you wrap the unsafe usage somewhere inside the VM?